exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableHandler.java

```diff
@@ -55,35 +58,79 @@ public DropTableHandler(SqlHandlerConfig config) {
+        } else {
+          throw UserException.validationError().message("Table [%s] not found", originalTableName).build(logger);
+        }
+      }
 
-    if (tableIdentifier != null) {
-      drillSchema = SchemaUtilites.resolveToMutableDrillSchema(defaultSchema, dropTableNode.getSchema());
+      try {
+        drillSchema.dropTable(originalTableName);
+      } catch (Exception e) {
+        // concurrency check: we might have failed because somebody has already dropped the table
+        if (SqlHandlerUtil.getTableFromSchema(drillSchema, originalTableName) != null) {
+          throw e;
+        }
+      }
     }
 
-    String tableName = dropTableNode.getName();
-    if (drillSchema == null) {
-      throw UserException.validationError()
-          .message("Invalid table_name [%s]", tableName)
-          .build(logger);
+    String message = String.format("%s [%s] dropped",
+        removedTemporaryTable ? "Temporary table" : "Table", originalTableName);
+    logger.info(message);
+    return DirectPlan.createDirectPlan(context, true, message);
+  }
+
+  /**
+   * If used workspace is default temporary workspace and temporary table is found,
+   * drops the table and removes it from session temporary tables list.
+   *
+   * @param defaultSchema default schema
+   * @param tableSchema table schema indicated in drop statement
+   * @param tableName   table name to drop
+   * @return true if temporary table existed and was dropped, false otherwise
+   */
+  private boolean removeTemporaryTable(SchemaPlus defaultSchema, List<String> tableSchema, String tableName) {
+    String temporaryWorkspace = context.getConfig().getString(ExecConstants.DEFAULT_TEMPORARY_WORKSPACE);
+
+    // look for temporary table only if default temporary workspace is used
+    if (!(tableSchema.isEmpty() || SchemaUtilites.getSchemaPath(tableSchema).equals(temporaryWorkspace))) {
+      return false;
     }
 
-    if (dropTableNode.checkTableExistence()) {
-      final Table tableToDrop = SqlHandlerUtil.getTableFromSchema(drillSchema, tableName);
-      if (tableToDrop == null || tableToDrop.getJdbcTableType() != Schema.TableType.TABLE) {
-        return DirectPlan.createDirectPlan(context, true,
-            String.format("Table [%s] not found", tableName));
-      }
+    String temporaryTable = context.getSession().findTemporaryTable(tableName);
+    if (temporaryTable == null) {
+      return false;
     }
 
-    drillSchema.dropTable(tableName);
+    AbstractSchema drillSchema = SchemaUtilites.resolveToMutableDrillSchema(defaultSchema,
+        Lists.newArrayList(temporaryWorkspace));
 
-    return DirectPlan.createDirectPlan(context, true,
-        String.format("Table [%s] %s", tableName, "dropped"));
+    Table tableToDrop = SqlHandlerUtil.getTableFromSchema(drillSchema, temporaryTable);
+    if (tableToDrop != null && tableToDrop.getJdbcTableType() == Schema.TableType.TABLE) {
```

Comment:
This logic looks very similar to that earlier for non-temp tables. Can it be factored out?